### PR TITLE
Docs: Explain TFTP and internal database

### PR DESCRIPTION
**Pull request by SchoolGuy**
@tomschr I would love an informal review as well if time allows. :)

This should clear up some confusion about topics often asked by admins and/or developers.

### CODE REVIEW

**Comment by olafhering** [comment]
In case the directories in `/srv/tftpboot` are really created on demand by cobblerd, they should be removed from the rpm filelist (remove `%{tftpboot_dir}/*` and related entries from `setup.py`.
This would be a followup fix.

The included `ls` output was done without locale `C`, maybe redo it with `env -i /bin/ls -l`.

File: docs/user-guide/tftp.rst [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+This directory will contain always the kernel and initrd of the distro you add to Cobbler. During a ``cobbler sync`` all
+folder with distros will be deleted and the structure will be recreated by the paths saved in the ``kernel`` and
+``initrd`` attributes in a Cobbler distro item.
+
+tftpboot/ipxe/default.ipxe
+==========================
+
+Cobbler will generate the iPXE menu for you. This is the file where all menu entries will be stored. It will be
+overwritten regularly by either a change in a distro or a ``cobbler sync``.
```
Comment (tomschr) [inline docs/user-guide/tftp.rst:108]:
```suggestion
overwritten regularly by either a change in a distro or by the command ``cobbler sync``.
```

**Review by tomschr — CHANGES_REQUESTED** [review]
@SchoolGuy: Wow, lots of work and effort you put into that. :+1: Thanks!
I have a couple of suggestions, ideas, and questions. Hope it makes sense. Try to remove "please", please. :slightly_smiling_face:

File: docs/user-guide/internal-database.rst [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+The internal database of Cobbler is held at ``/var/lib/cobbler/collections``.
+
+.. note:: There is an option to use a MongoDB but this is untested at the moment. The content of each entry is the same
+          as when using the default serializer.
+
+Items
+#####
+
+An Item in Cobbler is a set of attributes grouped together and given a name. An example for this would be a ``distro``.
+On disk those items are represented using JSON. Per default the JSON is minified, however you can make the serializer
```
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply docs/user-guide/internal-database.rst:16]: Yes I mean something like this.

File: docs/limitations.rst [hunk 3]
```diff
@@ -26,3 +26,10 @@ The Linux Kernel describes its quoting at:
 `The kernel’s command-line parameters <https://www.kernel.org/doc/html/v5.15/admin-guide/kernel-parameters.html#the-kernel-s-command-line-parameters>`_
 
 Please consult the documentation of your operating system for how it deals with this if it is not Linux.
+
+Special Case: Uyuni/SUSE Manager
+################################
+
+Uyuni and SUSE Manager don't see Cobbler as a source of truth and thus delete content they created right after they
+don't require Cobbler anymore again. Thus you see distros, profiles and systems popping on and off during the runtime
+of the server.
```
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply docs/limitations.rst]: "they" refers to both SUSE Manager and Uyuni in this case. Do you have a better suggestion in this case?

File: docs/user-guide/internal-database.rst [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+Notes
+#####
+
+Currently items get deleted upon deletion inside Cobbler immediately. If you want to have a backup please use the
+``scm_track`` module of Cobbler. It will use Git for version control of the complete ``/var/lib/cobbler/`` folder.
+
+A rename operation is currently: Delete the item with the old name and create a new item with the new name. This is
+reflected on disk and thus if your Cobbler is being terminated at the right point in time, there is the extremely low
+probability that this specific item is lost. If you have items dependent onto that item you will receive errors on the
```
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply docs/user-guide/internal-database.rst:34]: I do want to document that we have this edge-case in the database. Under normal operation it is not even a bug because it is a valid approach to a rename. However one would not expect this behavior when facing this particular situation and thus the sentence. I will apply your suggestion.

File: docs/user-guide/repository-management.rst [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+.. code-block:: shell
+
+    cobbler reposync --only="reponame1" ...
+
+When updating repos by name, a repo will be updated even if it is set to be not updated during a regular reposync
+operation (ex: ``cobbler repo edit --name=reponame1 --keep-updated=False``).
+
+Note that if a Cobbler import provides enough information to use the boot server as a yum mirror for core packages,
+Cobbler can set up automatic installation files to use the Cobbler server as a mirror instead of the outside world. If
```
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply docs/user-guide/repository-management.rst]: I did try to fix this with 160a62c

The commit messages will be improved once we are done with reviewing.

File: docs/user-guide/tftp.rst [hunk 6]
```diff
@@ -0,0 +1,124 @@
+******************
+The TFTP Directory
+******************
+
+For booting machines in a PXE and/or HTTP-Boot environment the TFTP directory is the most important directory. In this
+folder all static files required for booting are living.
```
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply docs/user-guide/tftp.rst]: Fixed via 1027877

File: docs/user-guide/tftp.rst [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+The folder of this is dependant on your distro and can be changed in the Cobbler settings. The default should be
+correctly set during the package build of your Linux distro or during the installation process (if you are use the
+source installation).
+
+Behaviour
```
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply docs/user-guide/tftp.rst:12]: I don't have the opinion that I am describing a workflow or procedure here. The code does not directly follow what I have written here. Thus in my eyes this is what the user should see if he watches the folder for changes. Thus this is the externally visible behaviour of the application in my eyes. Please try to explain further if you believe my argumentation is flawed.

File: docs/user-guide/tftp.rst [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    -rw-r--r--  1 root   root  54K Mär 17  2021 menu.c32
+    drwxr-xr-x  2 root   root   24 Dez 11  2020 others
+    lrwxrwxrwx  1 ftp    ftp    21 Dez  8  2020 p -> auto/dist/install/SLP
+    drwxr-xr-x  2 root   root   31 Nov 18 13:23 ppc
+    -rw-r--r--  1 root   root  26K Mär 17  2021 pxelinux.0
+    drwxr-xr-x  2 root   root  20K Jan 17 13:02 pxelinux.cfg
+    drwxr-xr-x  2 root   root 4,0K Nov 18 13:23 s390x
+
+
+All files or folders not covered by below explanations are specific to the environment the directory listing was taken
```
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply docs/user-guide/tftp.rst:74]: I wanted to supply a real example and not "construct" something. But I guess you are right and I can remove that listing completely.

File: docs/user-guide/tftp.rst [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+from. Those files should not be touched by Cobbler and should survive even a ``cobbler sync``.
+
+tftpboot/grub/
```
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply docs/user-guide/tftp.rst:77]: This is imitating a directory listing. I would very much like to keep this section.

File: docs/limitations.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (tomschr) [reply docs/limitations.rst]: Ahh I see! In that case, it's better to be explicit (although you need to repeat yourself). Also you may need to reorder it to apply the cause -> effect logic:

    If Uyuni and SUSE Manager don't require Cobbler anymore, Cobbler isn't seen
    as a source of truth. Thus, Uyuni and SUSE Manager delete content they have
    created which lead to popping on and off distros, profiles, and systems during
    the runtime of the server.

Does that make sense?

File: docs/user-guide/repository-management.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (tomschr) [reply docs/user-guide/repository-management.rst]: Much better! :+1:

File: docs/user-guide/tftp.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (tomschr) [reply docs/user-guide/tftp.rst:12]: Completely fine. Maybe I went a bit overboard and interpreted a bit too into that word. "Behaviour" reminds me a more of some human term rather than some machine term. Although it can be also the latter as I learned. Just ignore my question. :slightly_smiling_face:

File: docs/user-guide/tftp.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (tomschr) [reply docs/user-guide/tftp.rst]: I was not talking about removing the list completely. :wink:  More of reducing it to the minimum, what's really important. You still can say "This list is an excerpt. You will find more files inside this directory, however the following list covers only the most important ones." (something like that).

File: docs/user-guide/tftp.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (tomschr) [reply docs/user-guide/tftp.rst:43]: Yes, my point was to keep the listing and the explanation _together_ inside a list. Splitting it into different sections means, the user has (probably) to scroll back and forth.

You can keep it if you like it this way. :slightly_smiling_face:

File: docs/limitations.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply docs/limitations.rst]: Fixed via 569860c

File: docs/user-guide/repository-management.rst [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+
+Make sure there is plenty of space in Cobbler's webdir, which defaults to ``/var/www/cobbler``.
+
+.. code-block:: shell
+
+    cobbler reposync [--only=ONLY] [--tries=N] [--no-fail]
+
+Cobbler reposync is the command to use to update repos as configured with "cobbler repo add".  Mirroring
+can take a long time, and usage of Cobbler reposync prior to usage is needed to ensure provisioned systems have the
+files they need to actually use the mirrored repositories.  If you just add repos and never run "cobbler reposync", the
```
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply docs/user-guide/repository-management.rst]: Should be much better with a17a58b

File: docs/user-guide/tftp.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply docs/user-guide/tftp.rst:12]: Okay :)

File: docs/user-guide/tftp.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply docs/user-guide/tftp.rst:43]: Ahh you mean I shouldn't use headings. That is indeed a nice solution.

File: docs/user-guide/tftp.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply docs/user-guide/tftp.rst:75]: Okay will do. Thanks!

File: docs/user-guide/tftp.rst [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+#. The content of in above mentioned directories is being fully deleted.
+#. All bootloaders are being copied
+#. All kernel and initrds are being copied
+#. All images (if created) are being copied
+#. The PXE menu is being generated and written to disk
+#. The post-sync triggers are being executed
+
+.. note:: If you only sync DHCP, DNS or specific systems the order and actions might be slightly different.
+
+.. warning:: A ``cobbler sync`` is not required. Due to the file copying of a lot of small files this is a very
```
Comment (tomschr) [inline docs/user-guide/tftp.rst:36]: Just as a reminder: if possible, try to avoid consecutive admonitions. It make them less significant. But for the time being, leave it as.

**Review by tomschr — CHANGES_REQUESTED** [review]
@SchoolGuy Good job! Just some ideas, reminders, suggestions etc. Hope it helps.

File: docs/user-guide/tftp.rst (hunk 11 above)
Comment (SchoolGuy) [reply docs/user-guide/tftp.rst:36]: Thank your for the suggestion. Will try to avoid this in the future.

File: docs/user-guide/repository-management.rst [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+
+This example shows:
+
+ * How to create a provisioning infrastructure from a distribution mirror or from ISO media. 
+ * Create a default PXE configuration, so that by default systems will PXE boot into a fully automated install process for that distribution.
+You can use a network rsync mirror, a mounted DVD location, or a tree you have available via a network filesystem.
+
+Import knows how to autodetect the architecture of what is being imported. To make sure things are named
+correctly, it's a good idea to specify ``--arch``. For instance, if you import a distribution named "fedora8"
+from an x86_64 ISO, specify ``--arch=x86_64`` and the distro will be named "fedora8-x86_64"
```
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply docs/user-guide/repository-management.rst:127]: The repo part was only moved and comes from the days where Cobbler was maintained by RedHat, so I guess I need to change that a bit in the future... For now leaving as is due to time. :)

File: docs/user-guide/internal-database.rst [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+An item in Cobbler is a set of attributes grouped together and given a name. An example for this would be a ``distro``.
+On disk those items are represented using JSON. By default, the JSON is minified, however you can make the serializer
+produce "pretty" JSON files by changing ``serializer_pretty_json`` to ``true`` in the Cobbler Settings.
+
+The name of the saved file is the name of the item. However, this will be subject to change in the future.
+
+Collections
+###########
+
+A collection in Cobbler is a number of ``n`` Cobbler Items that are living inside the same folder.
```
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply docs/user-guide/internal-database.rst:24]: Indeed it is. Still lowercase is more intuitive to read and it is only an abstract concept. Thus applying your suggestion.

File: docs/user-guide/repository-management.rst [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+
+General
+#######
+
+This has already been covered a good bit in the command reference section, for details see: :ref:`cobbler-cli-reposync`
+
+Yum repository management is an optional feature and is not required to provision through Cobbler. However, if Cobbler
+is configured to mirror certain repositories, this feature can be used to associate profiles with those repositories.
+Systems installed under those profiles will be autoconfigured to use these repository mirrors in ``/etc/yum.repos.d``,
+and if supported (Fedora Core 6 and later), these repositories can be leveraged within Anaconda.
```
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply docs/user-guide/repository-management.rst:13]: Due to time I will leave it like it is. I will open an issue that this section is outdated and needs to be revised in a future release. Initially this was just a move of this part of the documentation.

File: docs/user-guide/repository-management.rst [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+desirable, it can be turned on by setting ``yum_post_install_mirror`` to ``True`` in ``/etc/cobbler/settings.yaml`` (and
+running ``cobbler sync``). You should not use this feature if machines are provisioned on a different VLAN/network than
+production, or if you are provisioning laptops that will want to acquire updates on multiple networks.
+
+The flags ``--tries=N`` (for example, ``--tries=3``) and ``--no-fail`` should likely be used when putting reposync on a
+crontab. They ensure network glitches in one repo can be retried and also that a failure to synchronize one repo does
+not stop other repositories from being synchronized.
+
+Importing trees
+###############
```
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply docs/user-guide/repository-management.rst:58]: You are right. Applying.

File: docs/user-guide/internal-database.rst [hunk 16]
```diff
@@ -0,0 +1,39 @@
+*****************
+Internal Database
+*****************
+
+.. note:: This document describes advanced topics for system administrators.
+
+The internal database of Cobbler is held at ``/var/lib/cobbler/collections``.
+
+.. note:: There is an option to use a MongoDB but this is untested at the moment. The content of each entry is the same
+          as when using the default serializer.
```
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply docs/user-guide/internal-database.rst:10]: Okay. My intention was to give the users a fair warning but you are probably right. Thus I will remove that note.

File: docs/user-guide/internal-database.rst [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+
+Items
+#####
+
+An item in Cobbler is a set of attributes grouped together and given a name. An example for this would be a ``distro``.
```
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply docs/user-guide/internal-database.rst:15]: `distro` is the name of a specific type of a Cobbler data-structure. Thus I will leave it like this. Everyone who reads this page should be aware of that and thus I believe although you are right I don't see an option to improve the situation at this moment.

File: docs/user-guide/repository-management.rst [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+
+#. you have a large install base,
+#. you want fast installation and upgrades for your systems, or
+#. have some extra software not in a standard repository but want provisioned systems to know about that repository.
+
+Make sure there is plenty of space in Cobbler's webdir, which defaults to ``/var/www/cobbler``.
+
+.. code-block:: shell
+
+    cobbler reposync [--only=ONLY] [--tries=N] [--no-fail]
```
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply docs/user-guide/repository-management.rst:25]: The page needs more care in a separate PR. Thus leaving as is for the moment.

File: docs/user-guide/repository-management.rst [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+
+For those familiar with dnf's reposync, Cobbler's reposync is mostly a wrapper around the ``dnf reposync`` command.
+use "cobbler reposync" to update Cobbler mirrors, as dnf's reposync does not perform all required steps. Also Cobbler
+adds support for rsync and SSH locations, where as dnf's reposync only supports what yum supports (http/ftp).
+
+If you want to update a certain repository, run:
+
+.. code-block:: shell
+
+    cobbler reposync --only="reponame1" ...
```
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply docs/user-guide/repository-management.rst:42]: The page needs more care in a separate PR. Thus leaving as is for the moment.

**Comment by SchoolGuy** [comment]
I will clean the Git History after the review.

File: docs/limitations.rst [hunk 20]
```diff
@@ -3,26 +3,35 @@ Limitations and Surprises
...
+Consult the documentation of your operating system for how it deals with this if it is not Linux.
+
+Special Case: Uyuni/SUSE Manager
+################################
+
+.. note:: SUSE Manager is a flavor of Uyuni. The term Uyuni refers to both pieces of software in this context.
+
+If Uyuni temporarily doesn't require Cobbler anymore, the content visible in Cobbler is deleted. Cobbler is thus not
+seen as a source of truth from the perspective of Uyuni. During the runtime of the application you may see distros,
+profiles, and systems popping on and off during the runtime of the server.
```
Comment (meaksh) [inline docs/limitations.rst:37]: I don't really get what you mean with this, don't really get the point.

In which cases "Uyuni temporarily doesn't require Cobbler anymore"?

Maybe I am missing context here.. Thanks!

File: docs/user-guide/repository-management.rst [hunk 21]
```diff
@@ -0,0 +1,144 @@
...
+production, or if you are provisioning laptops that will want to acquire updates on multiple networks.
+
+The flags ``--tries=N`` (for example, ``--tries=3``) and ``--no-fail`` should likely be used when putting reposync on a
+crontab. They ensure network glitches in one repo can be retried and also that a failure to synchronize one repo does
+not stop other repositories from being synchronized.
+
+Importing trees workflow
+########################
+
+Cobbler can auto-add distributions and profiles from remote sources, whether this is a filesystem path or an rsync
```
Comment (meaksh) [inline docs/user-guide/repository-management.rst:60]: Small typo:
```suggestion
Cobbler can auto-add distributions and profiles from remote sources, whether this is a filesystem path or a rsync
```

**Review by meaksh — COMMENTED** [review]
Thanks Enno, this looks good. I just added two minor comments :+1:

File: docs/limitations.rst (hunk 20 above)
Comment (agraul) [reply docs/limitations.rst:37]: I also don't get it, I think it would be good to clarify it a bit.

File: docs/user-guide/repository-management.rst (hunk 21 above)
Comment (agraul) [reply docs/user-guide/repository-management.rst]: As far as I understand the [grammar rules](https://www.merriam-webster.com/words-at-play/is-it-a-or-an), "an" is correct here. "A rsync" is hard to pronounce, at least for me.

File: docs/limitations.rst (hunk 20 above)
Comment (SchoolGuy) [reply docs/limitations.rst:37]: Okay this is slightly incorrect. It is only the systems are popping on and off. The reason for this is the taskomatic task `kickstart_cleanup`.

Comment (SchoolGuy) [reply docs/limitations.rst:37]: I will correct the note.

File: docs/user-guide/repository-management.rst (hunk 21 above)
Comment (SchoolGuy) [reply docs/user-guide/repository-management.rst]: Ahahaha okay I will revert it.

File: docs/limitations.rst (hunk 20 above)
Comment (meaksh) [reply docs/limitations.rst]: I still do not get that this means:

```
If Uyuni temporarily doesn't require Cobbler anymore, the content visible in Cobbler is deleted. Cobbler is thus not
seen as a source of truth from the perspective of Uyuni.
```

Could you maybe rephrase those two sentences? Particularly clarifying how "Uyuni can temporarily doesn't require Cobbler anymore"

Comment (SchoolGuy) [reply docs/limitations.rst]: Well if we don't do autoinstallations Cobbler is not used and thus not needed. Since Uyuni does not see this is a source of truth, it removes the content. I don't know how to phrase it better. Suggestions welcome!

Comment (meaksh) [reply docs/limitations.rst]: Ok, now you mentioned Autoinstallations, I think got your point! Thanks :)

I would maybe suggest something like this:

_Uyuni uses cobbler for driving autoinstallations. When using cobbler in the context of Uyuni, you need to know that cobbler is not seen by Uyuni as the source of truth. This means, in case you don't have any autoinstallation configured in Uyuni, the content visible in cobbler is deleted.
Because of the same reason, during the runtime of cobbler you may see systems popping on and off as the content of cobbler is managed by Uyuni (in particular, the taskomatic task `kickstart_cleanup` executes cleanup on the cobbler content)_

I think this brings a little bit more context to understand the point you want to indicate. What do you think?

Comment (SchoolGuy) [reply docs/limitations.rst]: Will apply this. If you think this is better then I believe it is. I am to deep in this. :d

**Review by meaksh — APPROVED** [review]

**Comment by SchoolGuy** [comment]
@meaksh @agraul @tomschr Thanks a lot for the reviews and discussions! Git History was cleaned. Thus I will merge this now!
